include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3312 @@
+    //               )
+    // After reduction:
+    //      uf_regs_dom = { k == t1; w == t2; v == t4; m == t2; }
+    //      base_dom = { 3 <= a; k <= w; w == m; }
+    //      Object = (
+    //                cache_dom = { z = 3, x <= y },
+    //                uf_flds_dom = { x == t1 ; y == t2; }
+    //               )
+    CRAB_LOG("object-reduce", crab::outs()
+                                  << "Before Reduction from base to cache:\n"
+                                  << "base = " << base_dom << "\n"
+                                  << "uf_regs = " << uf_regs_dom << "\n"
+                                  << "odi = ";
+             odi_write(crab::outs(), prod); crab::outs() << "\n";);
+    const uf_fields_abstract_domain_t &uf_flds_dom = prod.second().second();
+    base_abstract_domain_t &cache_dom = prod.second().first();
+    variable_vector_t remained_regs;
+    variable_vector_t renamed_flds;
+
+    symb_flds_regs_map_t map;
+    build_map(map, uf_regs_dom, uf_flds_dom, obj_flds);
+
+    // Note that, the regs-flds equalities are one-to-many
+    // A field could equal to multiple registers but
+    // multiple fields cannot equal to a single register
+
+    base_abstract_domain_t reduced_base = base_dom;
+
+    for (auto it = map.begin(); it != map.end(); ++it) {
+      const base_dom_variable_vector_t flds = std::get<0>(it->second);
+      const base_dom_variable_vector_t regs = std::get<1>(it->second);
+      assert(flds.size() > 0);
+      assert(regs.size() > 0);
+      const base_dom_variable_t &reg = regs[0];
+      if (flds.size() == 1) {
+        // reduced_base += ikos::operator==(reg, flds[0]);
+        reduced_base.rename({reg}, {flds[0]});
+      } else { // flds.size() > 1
+        auto it_flds = flds.begin();
+        const base_dom_variable_t &var = *it_flds;
+        while (it_flds != flds.end()) {
+          if (it_flds == flds.begin()) {
+            reduced_base.rename({reg}, {var});
+            // reduced_base += ikos::operator==(var, reg);
+          } else {
+            reduced_base += ikos::operator==(var, *it_flds);
+          }
+          ++it_flds;
+        }
+      }
+    }
+    // reduced_base.project(obj_flds);
+    reduced_base.project(obj_flds);
+    cache_dom = cache_dom & reduced_base;
+    CRAB_LOG("object-reduce", crab::outs()
+                                  << "After Reduction:\n"
+                                  << "base = " << base_dom << "\n"
+                                  << "uf_regs = " << uf_regs_dom << "\n"
+                                  << "odi = ";
+             odi_write(crab::outs(), prod); crab::outs() << "\n";);
+  }
+
+  // commit cache contents into summary if cache is dirty
+  // reset cache:
+  //  (1) perform reduction by transfering regs-flds' constriants between cache
+  //  domain and base domain (2) clear UF field domain (3) clear cache domain
+  void
+  commit_cache_if_dirty(base_abstract_domain_t &base_dom,
```

Comment:
if the object is top, just reset the cache/do nothing currently there is a lot of renaming/euf domain reasoning that could be fruitless